File: EVService/src/main/java/edu/uth/evservice/services/impl/ServiceItemService.java
package edu.uth.evservice.services.impl;

import edu.uth.evservice.dtos.ServiceItemDto;
import edu.uth.evservice.exception.ResourceNotFoundException;
import edu.uth.evservice.models.Part;
import edu.uth.evservice.models.ServiceItem;
import edu.uth.evservice.models.ServiceItemPart;
import edu.uth.evservice.models.ServiceItemPartId;
import edu.uth.evservice.repositories.IPartRepository;
import edu.uth.evservice.repositories.IServiceItemPartRepository;
import edu.uth.evservice.repositories.IServiceItemRepository;
import edu.uth.evservice.requests.ServiceItemRequest;
import edu.uth.evservice.services.IServiceItemService;
import lombok.*;
import lombok.experimental.FieldDefaults;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@AllArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
public class ServiceItemService implements IServiceItemService {

    IServiceItemRepository serviceItemRepository;
    IPartRepository partRepository;
    private final IServiceItemPartRepository suggestionRepo;

    @Override
    public ServiceItemDto addSuggestion(Integer itemId, Integer partId, int quantity) {
        ServiceItem item = serviceItemRepository.findById(itemId).orElseThrow(()-> new ResourceNotFoundException("Không tìm thấy hạng mục dịch vụ này")) ;

        Part part = partRepository.findById(partId).orElseThrow(() -> new ResourceNotFoundException("Không tìm thấy phụ tùng này"));

        ServiceItemPart suggestion = ServiceItemPart.builder()
                .id(new ServiceItemPartId(itemId,partId))
                .serviceItem(item)
                .part(part)
                .quantity(quantity)
                .build();
        suggestionRepo.save(suggestion);
        return toDto(item);
    }

    @Override
    public void removeSuggestion(Integer itemId, Integer partId) {
        ServiceItemPartId id = new ServiceItemPartId(itemId, partId);
        if (!suggestionRepo.existsById(id)) {
            throw new ResourceNotFoundException("Gợi ý không tồn tại");
        }
        suggestionRepo.deleteById(id);
    }

    @Override
    public List<ServiceItemDto> getAllServiceItems() {
        return serviceItemRepository.findAll()
                .stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    @Override
    public ServiceItemDto getServiceItemById(Integer id) {
        return serviceItemRepository.findById(id)
                .map(this::toDto)
                .orElse(null);
    }

    @Override
    public ServiceItemDto createServiceItem(ServiceItemRequest request) {
        ServiceItem item = new ServiceItem();
        item.setItemName(request.getItemName());
        item.setDescription(request.getDescription());
        item.setPrice(request.getPrice());

        ServiceItem saved = serviceItemRepository.save(item);
        return toDto(saved);
    }

    @Override
    public ServiceItemDto updateServiceItem(Integer id, ServiceItemRequest request) {
        return serviceItemRepository.findById(id)
                .map(existing -> {
                    existing.setItemName(request.getItemName());
                    existing.setDescription(request.getDescription());
                    existing.setPrice(request.getPrice());
                    ServiceItem updated = serviceItemRepository.save(existing);
                    return toDto(updated);
                })
                .orElse(null);
    }

    @Override
    public void deleteServiceItem(Integer id) {
        serviceItemRepository.deleteById(id);
    }

    private ServiceItemDto toDto(ServiceItem item) {
        ServiceItemDto dto = new ServiceItemDto();
        dto.setId(item.getItemId());
        dto.setItemName(item.getItemName());
        dto.setDescription(item.getDescription());
        dto.setPrice(item.getPrice());
        return dto;
    }
}

File: EVService-FE-Ts/src/services/serviceItemService.ts
// src/services/serviceItemService.ts

import api from "./api"; // Giả định 'api' là instance axios đã được cấu hình

// Tương đương ServiceItemDto
export interface SuggestedPart {
    partId: number;
    partName: string;
    quantity: number;
}

export interface ServiceItem {
    id: number;
    itemName: string;
    description: string;
    price: number;
    // Thêm trường này nếu ServiceItemDto của bạn trả về gợi ý
    suggestedParts: SuggestedPart[];
}

// Tương đương ServiceItemRequest
export interface ServiceItemRequest {
    itemName: string;
    description: string;
    price: number;
}

// Tương đương AddSuggestedPartRequest
export interface AddSuggestedPartRequest {
    partId: number;
    quantity: number;
}

export const serviceItemService = {
    // Lấy danh sách dịch vụ
    getAllServiceItems: async (): Promise<ServiceItem[]> => {
        const response = await api.get("/api/service-items");
        // Giả sử API trả về mảng ServiceItem, nếu không có suggestedParts thì gán mảng rỗng
        return response.data.map((item: any) => ({
            ...item,
            suggestedParts: item.suggestedParts || [],
        }));
    },

    // Tạo mới dịch vụ (ADMIN)
    createServiceItem: async (request: ServiceItemRequest): Promise<ServiceItem> => {
        const response = await api.post("/api/service-items", request);
        return response.data;
    },

    // Cập nhật dịch vụ (ADMIN)
    updateServiceItem: async (itemId: number, request: ServiceItemRequest): Promise<ServiceItem> => {
        const response = await api.put(`/api/service-items/${itemId}`, request);
        return response.data;
    },

    // Xóa dịch vụ (ADMIN)
    deleteServiceItem: async (itemId: number): Promise<void> => {
        await api.delete(`/api/service-items/${itemId}`);
    },

    // Thêm phụ tùng gợi ý (ADMIN)
    addSuggestedPart: async (itemId: number, request: AddSuggestedPartRequest): Promise<ServiceItem> => {
        const response = await api.post(`/api/service-items/${itemId}/suggest-part`, request);
        return response.data;
    },

    // Xóa phụ tùng gợi ý (ADMIN)
    removeSuggestedPart: async (itemId: number, partId: number): Promise<void> => {
        await api.delete(`/api/service-items/${itemId}/suggest-part/${partId}`);
    }
};